sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala

```diff
@@ -119,13 +130,66 @@ class KeyValueGroupedDataset[K, V] private[sql](
         Project(groupingAttributes, logicalPlan)))
   }
 
+  /**
+   * Returns a new [[KeyValueGroupedDataset]] with each group sorted by the given expressions.
+   * Operations that provide an iterator that contains all of the elements in a group will
+   * then provide a sorted iterator (flatMapGroups, mapGroups, cogroup).
+   *
+   * This is not supported for streaming Datasets (mapGroupsWithState, flatMapGroupsWithState).
+   *
+   * @tparam S The type of the sort value. Must be encodable to Spark SQL types.
+   * @param sortBy A function that provides a sort value for each row.
+   * @param direction The sort direction.
+   *
+   * @since 3.4.0
+   */
+  def sortWithinGroups[S: Encoder](
+      sortBy: V => S, direction: SortDirection = Ascending): KeyValueGroupedDataset[K, V] = {
```

Comment:
I feel it's tricky to define the sort order with a class `S`. Shall we respect its `compareTo` method? There is no `Dataset.sortBy` with a class either. Shall we remove this overload?

Reply:
This was inspired by `groupByKey`. You are right, equality (`groupByKey`) and order are two different things. since there is no equivalent `Dataset.sortBy` method, I'll remove this.